flecsi/data/hpx/policy.hh

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2016, Triad National Security, LLC
+// All rights reserved.
+
+#ifndef FLECSI_DATA_HPX_POLICY_HH
+#define FLECSI_DATA_HPX_POLICY_HH
+
+#include <flecsi-config.h>
+
+#if !defined(FLECSI_ENABLE_HPX)
+#error FLECSI_ENABLE_HPX not defined! This file depends on HPX!
+#endif
+
+#include <hpx/modules/collectives.hpp>
+#include <hpx/modules/futures.hpp>
+
+#include "flecsi/data/local/storage.hh"
+#include "flecsi/exec/hpx/future.hh"
+#include "flecsi/run/hpx/context.hh"
+#include "flecsi/util/types.hh"
+
+#include <cstddef>
+#include <cstring>
+#include <string>
+#include <vector>
+
+namespace flecsi {
+namespace data {
+
+/// \defgroup hpx-data HPX Data
+/// Type of the dependency represented by a future
+/// \ingroup data
+/// \{
+enum class dependency : std::uint8_t {
+  /// This field has no future associated with it (yet)
+  none = 0,
+  /// The future represents a read operation
+  read = 1,
+  /// The future represents a write operation
+  write = 2,
+  /// The future represents the result of a ghost_copy operation
+  ghost_copy = 3
+};
+/// \}
+
+/// Direct data storage.
+/// \ingroup data
+/// \{
+struct backend_storage : local::detail::storage {
+
+  ~backend_storage() {
+    // whenever region is destroyed we have to wait for all pending tasks to
+    // complete
+    if(future.valid() && !future.is_ready()) {
+      future.wait();
+      if(future.has_exception()) {
+        // there is no way to report the error to the user's code at this point,
+        // thus termination is the only option
+        flog_fatal(
+          "future is in exceptional state during destruction of region");
+      }
+    }
+  }
+
+  ::hpx::shared_future<void> future;
+  dependency dep = dependency::none;
+};
+/// \}
+
+/// Synchronize with all pending operations on this storage.
+/// \ingroup data
+/// \{
+inline void synchronize(backend_storage& s)
```

Comment:
It seems like this wants to be a member function.

Reply:
I thought about that, but then decided against it to avoid having to create a `backend_storage` type for the MPI backend (it's an alias, currently). I will make the change, though.

Reply:
Sorry: I didn't realize that that function existed in both backends.  Perhaps we should have another `#ifdef DOXYGEN` block in `local/storage.hh` to describe that expected interface (even though it would appear only in the Core API reference)?

Reply:
Done.